runtime/src/parachains.rs

```diff
@@ -144,49 +177,119 @@ decl_storage! {
 
 decl_module! {
 	/// Parachains module.
-	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+	pub struct Module<T: Trait> for enum Call where origin: <T as system::Trait>::Origin {
 		/// Provide candidate receipts for parachains, in ascending order by id.
 		fn set_heads(origin, heads: Vec<AttestedCandidate>) -> Result {
 			ensure_none(origin)?;
 			ensure!(!<DidUpdate<T>>::exists(), "Parachain heads must be updated only once in the block");
 
 			let active_parachains = Self::active_parachains();
+			let parachain_count = active_parachains.len();
+			ensure!(heads.len() <= parachain_count, "Too many parachain candidates");
+
+			if !active_parachains.is_empty() {
+				// perform integrity checks before writing to storage.
+				{
+					let mut last_id = None;
+					let mut iter = active_parachains.iter();
+					for head in &heads {
+						let id = head.parachain_index();
+						// proposed heads must be ascending order by parachain ID without duplicate.
+						ensure!(
+							last_id.as_ref().map_or(true, |x| x < &id),
+							"Parachain candidates out of order by ID"
+						);
+
+						// must be unknown since active parachains are always sorted.
+						ensure!(
+							iter.find(|x| x == &&id).is_some(),
+							"Submitted candidate for unregistered or out-of-order parachain {}"
+						);
+
+						// Either there are no more messages to add...
+						if !head.candidate.upward_messages.is_empty() {
+							let (count, size) = <RelayDispatchQueueSize<T>>::get(id);
+							ensure!(
+								// ...or we are appending one message onto an empty queue...
+								head.candidate.upward_messages.len() + count as usize == 1
+								// ...or...
+								|| (
+									// ...the total messages in the queue ends up being no greater than the
+									// limit...
+									head.candidate.upward_messages.len() + count as usize <= MAX_QUEUE_COUNT
+								&&
+									// ...and the total size of the payloads in the queue ends up being no
+									// greater than the limit.
+									head.candidate.upward_messages.iter()
+										.fold(size as usize, |a, x| a + x.1.len())
```

Comment:
This could be subject to integer overflow.

Reply:
the messages are held in memory. `usize` is capable of expressing the size of memory. qed.